phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -999,6 +1007,23 @@ private void addCoprocessors(byte[] tableName, HTableDescriptor descriptor, PTab
         }
     }
 
+    private boolean doesPhoenixTableAlreadyExist(HTableDescriptor existingDesc) {
+        //if the table descriptor already has Phoenix coprocs, we assume it's
+        //already gone through a Phoenix create statement once
+        if (existingDesc == null){
+            return false;
+        }
+        boolean hasScanObserver = existingDesc.hasCoprocessor(ScanRegionObserver.class.getName());
+        boolean hasUnAggObserver = existingDesc.hasCoprocessor(
+            UngroupedAggregateRegionObserver.class.getName());
+        boolean hasGroupedObserver = existingDesc.hasCoprocessor(
+            GroupedAggregateRegionObserver.class.getName());
+        boolean hasIndexObserver = existingDesc.hasCoprocessor(Indexer.class.getName())
+            || existingDesc.hasCoprocessor(IndexRegionObserver.class.getName())
+            || existingDesc.hasCoprocessor(GlobalIndexChecker.class.getName());
+        return hasScanObserver && hasUnAggObserver && hasGroupedObserver && hasIndexObserver;
```

Comment:
Rather than checking coprocessors one by one, why don't we check that the coprocessors on the table length is greater than 0?

Reply:
Coprocs could exist for other, non-Phoenix reasons, or be incomplete because of some prior metadata corruption.

Reply:
If any of these coprocs are missing, we would still end-up adding the indexing coprocs, possibly under the similar rolling upgrade time as we have recently experienced. Should we confirm if 'CREATE .. IF NOT EXISTS' is supposed to so such 'repairs' or are there other better mechanisms?

Reply:
@priyankporwal - the tricky thing is that it's legal to take a raw HBase table, run a "CREATE TABLE" statement, and expect it to get the coprocs. Proving that that's not what's happening (or is in the middle of happening on some other transaction) is hard, so I'm being conservative + not trying to solve the general problem of making an existent table not touched (which I agree is the ideal behavior.)